src/main/java/net/imagej/ops/image/watershed/Watershed.java

```diff
@@ -0,0 +1,393 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+package net.imagej.ops.image.watershed;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+
+import net.imagej.ops.Contingent;
+import net.imagej.ops.Ops;
+import net.imagej.ops.create.imgLabeling.CreateImgLabelingFromInterval;
+import net.imagej.ops.special.function.Functions;
+import net.imagej.ops.special.function.UnaryFunctionOp;
+import net.imagej.ops.special.hybrid.AbstractUnaryHybridCF;
+import net.imglib2.Cursor;
+import net.imglib2.FinalInterval;
+import net.imglib2.RandomAccess;
+import net.imglib2.RandomAccessible;
+import net.imglib2.RandomAccessibleInterval;
+import net.imglib2.algorithm.neighborhood.DiamondShape;
+import net.imglib2.algorithm.neighborhood.Neighborhood;
+import net.imglib2.algorithm.neighborhood.RectangleShape;
+import net.imglib2.algorithm.neighborhood.Shape;
+import net.imglib2.roi.labeling.ImgLabeling;
+import net.imglib2.roi.labeling.LabelingType;
+import net.imglib2.type.BooleanType;
+import net.imglib2.type.numeric.RealType;
+import net.imglib2.type.numeric.integer.IntType;
+import net.imglib2.util.IntervalIndexer;
+import net.imglib2.util.Intervals;
+import net.imglib2.view.Views;
+
+import org.scijava.plugin.Parameter;
+import org.scijava.plugin.Plugin;
+
+/**
+ * The Watershed algorithm segments and labels a grayscale image analogous to a
+ * heightmap. In short, a drop of water following the gradient of an image flows
+ * along a path to finally reach a local minimum.
+ *
+ * Lee Vincent, Pierre Soille, Watersheds in digital spaces: An efficient
+ * algorithm based on immersion simulations, IEEE Trans. Pattern Anal. Machine
+ * Intell., 13(6) 583-598 (1991)
+ *
+ * Input is a grayscale image with arbitrary number of dimensions, defining the
+ * heightmap. It needs to be defined whether a neighborhood with eight- or
+ * four-connectivity (respective to 2D) is used. A binary image can be set as
+ * mask which defines the area where computation shall be done.
+ *
+ * Output is a labeling of the different catchment basins.
+ *
+ * @author Simon Schmid (University of Konstanz)
+ */
+@Plugin(type = Ops.Image.Watershed.class)
+public class Watershed<B extends BooleanType<B>, T extends RealType<T>>
+		extends AbstractUnaryHybridCF<RandomAccessibleInterval<T>, ImgLabeling<Integer, IntType>>
+		implements Ops.Image.Watershed, Contingent {
+
+	@SuppressWarnings("rawtypes")
+	private UnaryFunctionOp<FinalInterval, ImgLabeling> createOp;
+
+	@Parameter(required = true)
+	boolean eightConnectivity;
+
+	@Parameter(required = false)
+	RandomAccessibleInterval<B> mask;
+
+	@SuppressWarnings("null")
+	@Override
+	public void compute1(final RandomAccessibleInterval<T> in, final ImgLabeling<Integer, IntType> out) {
```

Comment:
Ah ok, so you mean in the WatershedSeeded class. Advantage of using ImgLabeling for seeds is that the labels of the segments can be determined, because every segment will be labeled accoring to its seed.